File: gfg/POTD/2025-02-20/solution.js
/**
 * @param {number[]} arr
 * @returns {number[]}
 */

class Solution {
	getMedian(arr) {
		let result = []
		let up = [],
			dn = []

		for (let ve of arr) {
			up.push(ve)
			up.sort((a, b) => a - b)
			ve = up.shift()

			dn.push(-ve)
			dn.sort((a, b) => a - b)
			ve = -dn.shift()

			if (up.length === dn.length) {
				result.push(ve)
				up.push(ve)
			} else if (up.length > dn.length) {
				result.push((up[0] + ve) / 2)
				dn.push(-ve)
			} else {
				result.push((-dn[0] + ve) / 2)
				up.push(ve)
			}
		}
		return result
	}
}

File: gfg/POTD/2025-02-20/solution.java
class Solution {
    public ArrayList<Double> getMedian(int[] arr) {
        ArrayList<Double> result = new ArrayList<>();
        PriorityQueue<Integer> up = new PriorityQueue<>();
        PriorityQueue<Integer> dn = new PriorityQueue<>(Collections.reverseOrder());

        for (int ve : arr) {
            up.offer(ve);
            ve = up.poll();
            dn.offer(ve);
            ve = dn.poll();
            
            if (up.size() == dn.size()) {
                result.add((double) ve);
                up.offer(ve);
            } else if (up.size() > dn.size()) {
                result.add((up.peek() + (double) ve) / 2);
                dn.offer(ve);
            } else {
                result.add(((-dn.peek() + (double) ve) / 2));
                up.offer(ve);
            }
        }
        return result;
    }
}